File: 142-linked-list-cycle-ii/142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
   if(head==null||head.next==null) return null;
        
         ListNode sp=head;       
         ListNode fp=head;     
        
        do{
          sp=sp.next;
            fp=fp.next.next;
        }while(sp!=fp && fp!=null && fp.next!=null);
        
    
        if(sp==fp){ //loop detected
            sp=head;
            while(sp!=fp){
                sp=sp.next;
                fp=fp.next;
            }
            return sp;
        }
        return null;

    }
}

File: 142-linked-list-cycle-ii/142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next: return None
        sp=head.next
        fp=head.next.next
        
        while sp!=fp and fp and fp.next:
            sp=sp.next
            fp=fp.next.next
        if not fp or not fp.next: return None
        fp=head
        pos=0
        while fp!=sp:
            sp=sp.next
            fp=fp.next
            pos+=1
        return sp
        